### PR TITLE
ACN_CD_Netskope_DataConnector01

### PR DESCRIPTION
**Pull request by chicduong**
## Proposed Changes

  - Netskope custom data connector
  - Please provide link for
      a) Azure Deploy Button
      b) Kusto Function / Parser
  - After merge, subsequent PR will be submitted to to update
     a) Azure Deploy (json) zip package link
     b) Function App code link, on the connector page

### CODE REVIEW

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+# Get the current universal time in the default string format.
+$currentUTCtime = (Get-Date).ToUniversalTime()
+
+# The 'IsPastDue' property is 'true' when the current function invocation is later than scheduled.
+if ($Timer.IsPastDue) {
+    Write-Host "PowerShell timer is running late!"
+}
+
+# Function to contruct the Netskope Uri for alerts, event types, and to accomodate for pagination
+function GetUrl ($uri, $ApiKey, $StartTime, $EndTime, $LogType, $Page){
```
Comment (nazang) [inline DataConnectors/Netskope/AzureFunctionNetskope/run.ps1:26]: $skip needs to be passed in as a parameter as well?

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+            
+    foreach($logtype in $apitypes){
+
+        $endTime = (Get-Date -Date ((Get-Date).DateTime) -UFormat %s)  
+        $startTime = GetStartTime -CheckpointFile $checkPointFile -LogType $logtype -TimeInterval $timeInterval # function to create starttime
+
+        $alleventobjs = @()
+        $count = 0
+        Do {
+            $response = GetLogs -Uri $uri -ApiKey $apikey -StartTime $startTime -EndTime $endTime -LogType $logtype -Page $pageLimit
```
Comment (nazang) [inline DataConnectors/Netskope/AzureFunctionNetskope/run.ps1:112]: Related to previous comment, $skip needs to pass-in to GetLogs as a param

**Review by nazang — CHANGES_REQUESTED** [review]

File: DataConnectors/Netskope/Netskope_API_FunctionApp.json [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                "provider": "Microsoft.OperationalInsights/workspaces",
+                "permissionsDisplayText": "read and write permissions are required.",
+                "providerDisplayName": "Workspace",
+                "scope": "Workspace",
+                "requiredPermissions": {
+                    "write": true,
+                    "read": true,
+                    "delete": true
+                }
+            }
```
Comment (v-jayakal) [inline DataConnectors/Netskope/Netskope_API_FunctionApp.json:54]: Missing permissions - "Microsoft.OperationalInsights/workspaces/sharedKeys", Please add. Refer https://github.com/Azure/Azure-Sentinel/blob/master/DataConnectors/Templates/Connector_REST_API_AzureFunctionApp_template/DataConnector_API_AzureFunctionApp_template.json

**Review by v-jayakal — CHANGES_REQUESTED** [review]
Name of the sample data file should be data type name (Netskope_CL).

File: DataConnectors/Netskope/Netskope_API_FunctionApp.json [hunk 4]
```diff
@@ -0,0 +1,128 @@
+{
+    "id": "Netskope",
+    "title": "Netskope",
+    "publisher": "Netskope",
+    "descriptionMarkdown": "The [Netskope Cloud Security Platform](https://www.netskope.com/platform) connector provides the capability to ingest Netskope logs and events into Azure Sentinel. The connector provides visibility into Netskope Platform Events and Alerts in Azure Sentinel to improve monitoring and investigation capabilities.",
+    "additionalRequirementBanner":"These queries are based on a Kusto Function to work as expected. Follow the steps to use this Kusto functions alias **Netskope**. [Follow these steps to get this Kusto functions.](https://aka.ms/sentinelgithubparsersnetskope)",
```
Comment (nazang) [inline DataConnectors/Netskope/Netskope_API_FunctionApp.json:6]: @chicduong, we didn't see the parser PR or it is not included in this PR, is it missed?

**Review by nazang — CHANGES_REQUESTED** [review]

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 [hunk 5]
```diff
@@ -0,0 +1,198 @@
+<#  
+    Title:          Netskope Connector
+    Language:       PowerShell
+    Version:        1.0
+    Author(s):      Microsoft
+    Last Modified:  11/13/2020
+    Comment:        Initial Release
+
+    DESCRIPTION
+    This Function App calls the Netskope Platform API (https://innovatechcloud.goskope.com/docs/Netskope_Help/en/rest-api-v2-overview.html) to pull alert and events data. The response from the Okta API is recieved in JSON format. This function will build the signature and authorization header
```
[the remark this answers is not in the document]
Comment (chicduong) [reply DataConnectors/Netskope/AzureFunctionNetskope/run.ps1]: Yup, this is a typo. revised and submitted.

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 (hunk 1 above)
Comment (chicduong) [reply DataConnectors/Netskope/AzureFunctionNetskope/run.ps1]: this isn't necessary, but I have added for code comprehensibility

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 (hunk 2 above)
Comment (chicduong) [reply DataConnectors/Netskope/AzureFunctionNetskope/run.ps1]: this isn't necessary, but I have added for code comprehensibility

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    }
+    else{
+        return $url
+    }
+}
+
+
+# Function for retrieving alerts and events from Netskope's APIs
+function GetLogs ($Uri, $ApiKey, $StartTime, $EndTime, $LogType, $Page, $Skip){
+    $url = GetUrl -Uri $Uri -ApiKey $ApiKey -StartTime $StartTime -EndTime $EndTime -logtype $LogType -Page $Page -Skip $skip
```
Comment (nazang) [inline DataConnectors/Netskope/AzureFunctionNetskope/run.ps1:46]: Capital S  --> $Skip

File: DataConnectors/Netskope/Netskope_API_FunctionApp.json (hunk 4 above)
Comment (chicduong) [reply DataConnectors/Netskope/Netskope_API_FunctionApp.json:6]: my apologies, forgot to submit the PR after pushing changes. https://github.com/Azure/Azure-Sentinel/pull/1332

File: DataConnectors/Netskope/AzureFunctionNetskope/run.ps1 (hunk 6 above)
Comment (chicduong) [reply DataConnectors/Netskope/AzureFunctionNetskope/run.ps1]: updated and submitted

File: DataConnectors/Netskope/Netskope_API_FunctionApp.json (hunk 3 above)
Comment (chicduong) [reply DataConnectors/Netskope/Netskope_API_FunctionApp.json]: updated and submitted.

**Review by v-jayakal — APPROVED** [review]

**Review by erant10 — APPROVED** [review]

**Review by nazang — APPROVED** [review]

**Pull request merged**